### PR TITLE
Add function + tests to convert decimals with token

### PR DESCRIPTION
**Pull request by joYyHack**
- [ ] Since this PR suggests a **bug fix**, the tests have been added and the coverage is 100%.
- [x] Since this PR introduces a **new feature**, the update has been discussed in an Issue or with the team.
- [ ] This PR is just a **minor change**, like a typo fix.

---

Added token decimals converter. The function accepts:

1. The amount of the base token.
2. The base token address.
3. The destination token address.

Under the hood, the function casts the addresses of the tokens to the IERC20Metadata interface and retrieves the decimals for both tokens. Then, it converts the amount with the decimals of the base token to the decimals of the destination token.

All necessary tests are added. Coverage 100%.

### CODE REVIEW

File: contracts/libs/decimals/DecimalsConverter.sol [hunk 1]
```diff
@@ -104,6 +104,41 @@ library DecimalsConverter {
...
+    function roundTokens(
+        uint256 amount_,
+        address baseToken_,
+        address destToken_
+    ) internal view returns (uint256) {
+        return
+            convert(
+                amount_,
+                uint256(IERC20Metadata(baseToken_).decimals()),
+                uint256(IERC20Metadata(destToken_).decimals())
```
Comment (Hrom131) [inline contracts/libs/decimals/DecimalsConverter.sol:123]: We can use `decimals` function from the current library

**Comment by Hrom131** [comment]
Let's add functions `to18` and `from18`, which will take the address of the token, instead of decimals

File: contracts/libs/decimals/DecimalsConverter.sol [hunk 2]
```diff
@@ -104,6 +104,41 @@ library DecimalsConverter {
...
     }
 
+    /**
+     * @notice The function to do the token precision convertion
+     * @param amount_ the amount to convert
+     * @param baseToken_ current token
+     * @param destToken_ desired token
+     * @return the converted number
+     */
+    function roundTokens(
```
Comment (Arvolear) [inline contracts/libs/decimals/DecimalsConverter.sol:114]: Could you please explain why would one need this function?

**Comment by Arvolear** [comment]
> Let's add functions `to18` and `from18`, which will take the address of the token, instead of decimals

This is indeed what will be useful

File: contracts/libs/decimals/DecimalsConverter.sol (hunk 1 above)
Comment (joYyHack) [reply contracts/libs/decimals/DecimalsConverter.sol:123]: sure

File: contracts/libs/decimals/DecimalsConverter.sol (hunk 2 above)
Comment (joYyHack) [reply contracts/libs/decimals/DecimalsConverter.sol:114]: Well, I had a case in a previous project where I needed to compare the amount of stable coins from Curve, and I had to adjust one amount to match the decimal precision of the second amount. So, when I wrote this function, I was thinking about that particular case. 😄

**Comment by joYyHack** [comment]
> Let's add functions `to18` and `from18`, which will take the address of the token, instead of decimals

**Comment by joYyHack** [comment]
> > Let's add functions `to18` and `from18`, which will take the address of the token, instead of decimals
> 
> This is indeed what will be useful

Okay, that's a good idea, but I need to warn you that it appears impossible to overload 'safe' functions because they all utilize the 'convert' function, which accepts a function as a parameter. So, for example, if we have both 
'to18(uint256, uint256)' and 'to18(uint256, address)' along with 'convert(function(uint256, uint256)),' and we attempt to call 'function test { convert(to18) },' the compiler will report an error, stating 'No matching declaration found after variable lookup.' It seems that the compiler can't determine which of the overloaded functions we're trying to use.
image

Overall the idea is that I will just name it differently :D

File: contracts/libs/decimals/DecimalsConverter.sol [hunk 3]
```diff
@@ -150,7 +197,7 @@ library DecimalsConverter {
         uint256 amount_,
         uint256 baseDecimals_,
         uint256 destDecimals_
-    ) internal pure returns (uint256) {
+    ) private pure returns (uint256) {
```
Comment (Arvolear) [inline contracts/libs/decimals/DecimalsConverter.sol:200]: Please leave it `internal`

File: contracts/libs/decimals/DecimalsConverter.sol [hunk 4]
```diff
@@ -104,6 +145,47 @@ library DecimalsConverter {
...
     }
 
+    /**
+     * @notice The function to do the token precision convertion
+     * @param amount_ the amount to convert
+     * @param baseToken_ current token
+     * @param destToken_ desired token
+     * @return the converted number
+     */
+    function roundTokens(
```
Comment (Arvolear) [inline contracts/libs/decimals/DecimalsConverter.sol:155]: Let's name it `convert`

File: contracts/libs/decimals/DecimalsConverter.sol [hunk 5]
```diff
@@ -104,6 +145,47 @@ library DecimalsConverter {
...
+     */
+    function roundTokensSave(
+        uint256 amount_,
+        address baseToken_,
+        address destToken_
+    ) internal view returns (uint256) {
+        return convertTokensSafe(amount_, baseToken_, destToken_, roundTokens);
+    }
+
+    function precisionTo18(uint256 amount_, uint256 baseDecimals_) private pure returns (uint256) {
```
Comment (Arvolear) [inline contracts/libs/decimals/DecimalsConverter.sol:178]:
```suggestion
    function _precisionTo18(uint256 amount_, uint256 baseDecimals_) private pure returns (uint256) {
```

File: contracts/libs/decimals/DecimalsConverter.sol [hunk 6]
```diff
@@ -104,6 +145,47 @@ library DecimalsConverter {
...
+    }
+
+    /**
+     * @notice The function to do the token precision convertion. Reverts if output is zero
+     * @param amount_ the amount to convert
+     * @param baseToken_ current token
+     * @param destToken_ desired token
+     * @return the converted number
+     */
+    function roundTokensSave(
```
Comment (Arvolear) [inline contracts/libs/decimals/DecimalsConverter.sol:170]:
```suggestion
    function roundTokensSafe(
```

File: test/libs/decimals/DecimalsConverter.test.js [hunk 7]
```diff
@@ -100,4 +147,31 @@ describe("DecimalsConverter", () => {
...
+
+  describe("roundTokens", () => {
+    it("should round tokens decimals", async () => {
+      const token1 = await ERC20Mock.new("MK1", "MK1", 18);
+      const token2 = await ERC20Mock.new("MK2", "MK2", 3);
+
+      assert.equal((await mock.roundTokens(wei("1"), token1.address, token2.address)).toFixed(), wei("1", 3));
+      assert.equal((await mock.roundTokens(wei("1", 3), token2.address, token1.address)).toFixed(), wei("1"));
+    });
+  });
```
Comment (Arvolear) [inline test/libs/decimals/DecimalsConverter.test.js:159]:
```suggestion
  });
  
```

File: test/libs/decimals/DecimalsConverter.test.js [hunk 8]
```diff
@@ -100,4 +147,31 @@ describe("DecimalsConverter", () => {
...
+  describe("roundTokensSafe", () => {
+    it("should round tokens decimals", async () => {
+      const token1 = await ERC20Mock.new("MK1", "MK1", 6);
+      const token2 = await ERC20Mock.new("MK2", "MK2", 9);
+
+      assert.equal((await mock.roundTokensSave(wei("1", 6), token1.address, token2.address)).toFixed(), wei("1", 9));
+      assert.equal((await mock.roundTokensSave(wei("1", 9), token2.address, token1.address)).toFixed(), wei("1", 6));
+    });
```
Comment (Arvolear) [inline test/libs/decimals/DecimalsConverter.test.js:167]:
```suggestion
    });
    
```

File: test/libs/decimals/DecimalsConverter.test.js [hunk 9]
```diff
@@ -25,51 +25,98 @@ describe("DecimalsConverter", () => {
...
+      );
     });
   });
 
   describe("from18Safe", () => {
-    it("should correctly convert to 18", async () => {
+    it("should correctly convert from 18", async () => {
       assert.equal((await mock.from18Safe(wei("1"), 6)).toFixed(), wei("1", 6));
       assert.equal((await mock.from18Safe(wei("1"), 8)).toFixed(), wei("1", 8));
     });
```
Comment (Arvolear) [inline test/libs/decimals/DecimalsConverter.test.js:106]:
```suggestion
    });
    
```

**Pull request merged**